src/Vega4Wrapper.js

```diff
@@ -0,0 +1,207 @@
+     * @param {Function} wrapperOpts.logger
+     * @param {Function} wrapperOpts.formatUrl
+     * @param {string} [wrapperOpts.languageCode]
+     * @constructor
+     */
+    constructor(wrapperOpts) {
+        var self = this;
+        // Copy all options into the wrapper
+        Object.assign(self, wrapperOpts);
+        self.validators = {};
+
+        self.loader.sanitize = self.sanitize.bind(self);
+        self.loader.load = (uri, options) => {
+            return self.sanitize(uri, options)
+                .then(opt => {
+                    var url = opt.href;
+                    return self.loader.http(url, options)
+                        .then(txt => {
+                            return self.parseResponse(txt, uri.type, options);
+                        });
+                });
+        }
+
+        // Prevent accidental use
+        self.loader.file = () => { throw new Error('Disabled'); };
+    }
+
+    /**
+     * Validate and update urlObj to be safe for client-side and server-side usage
+     * @param {object} uri - An object that will be converted into an url string
+     * @param {object} options - passed by the vega loader
+     * @return {Promise} The sanitized url is provided by the 'href' property. We never load file from local file system.
+     */
+    sanitize(uri, options) {
+        var url = this.objToUrl(uri, options);
+        return new Promise(function(accept) {
+            var result = {href: url, loadFile: false};
+            
+            //return
+            accept(result);
+        });
+    }
+
+    /**
+     * Check if host was listed in the allowed domains, normalize it, and get correct protocol
+     * @param {string} host
+     * @returns {Object}
+     */
+    sanitizeHost(host) {
+        // First, map the host
+        host = (this.domainMap && this.domainMap[host]) || host;
+
+        if (this.testHost('https', host)) {
+            return { host: host, protocol: 'https' };
+        } else if (this.testHost('http', host)) {
+            return { host: host, protocol: 'http' };
+        }
+        return undefined;
+    };
+
+    /**
+     * Test host against the list of allowed domains based on the protocol
+     * @param {string} protocol
+     * @param {string} host
+     * @returns {boolean}
+     */
+    testHost(protocol, host) {
+        if (!this.validators[protocol]) {
+            var domains = this.domains[protocol];
+            if (domains) {
+                this.validators[protocol] = makeValidator(domains, protocol === 'https' || protocol === 'http');
+            } else {
+                return false;
+            }
+        }
+        return this.validators[protocol].test(host);
+    };
+    
+    /**
+     * convert a uri object to a url string
+     * @param {object} uri an object consists of type and essential parameters
+     * @param {object} options used to attach CORS infomation
+     * @returns {string} a complete url
+     */
+    objToUrl(uri, options) {
```

Comment:
uri is confusing - maybe `urlObj` ?    Also, `uri` is an ID, so we shouldn't use it regardless.